core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala

```diff
@@ -723,14 +726,14 @@ class GroupCoordinator(val brokerId: Int,
         responseCallback(offsetMetadata.map { case (k, _) => k -> Errors.COORDINATOR_NOT_AVAILABLE })
       } else if (group.isStaticMemberFenced(memberId, groupInstanceId)) {
         responseCallback(offsetMetadata.map { case (k, _) => k -> Errors.FENCED_INSTANCE_ID })
-      } else if ((generationId < 0 && group.is(Empty)) || (producerId != NO_PRODUCER_ID)) {
+      } else if (generationId >= 0 && generationId != group.generationId) {
+        // Validate non-zero generation id for both transactional and non-transactional commits.
+        responseCallback(offsetMetadata.map { case (k, _) => k -> Errors.ILLEGAL_GENERATION })
+      } else if (group.isUnknownCommit(memberId, producerId)) {
+        responseCallback(offsetMetadata.map { case (k, _) => k -> Errors.UNKNOWN_MEMBER_ID })
+      } else if (generationId < 0 && group.is(Empty) || producerId != NO_PRODUCER_ID) {
```

Comment:
It's debatable whether we should continue going to group state based for commit logic for transactional commit as well. We decide to leave it as it is for now, unless the group state becomes concerning factor.